mpd/asyncio.py

```diff
@@ -0,0 +1,406 @@
+                                self._write_command("noidle")
+                                self.__command_enqueued = None
+                            l = await line_future
+                        else:
+                            l = await self.__read_output_line()
+                    except CommandError as e:
+                        result._feed_error(e)
+                        break
+                    result._feed_line(l)
+                    if l is None:
+                        break
+
+                result = None
+
+        except Exception as e:
+            # Prevent the destruction of the pending task in the shutdown
+            # function -- it's just shutting down by itself.
+            self.__run_task = None
+            self.disconnect()
+
+            if result is not None:
+                result._feed_error(e)
+                return
+            else:
+                # Typically this is a bug in mpd.asyncio.
+                raise
+
+    async def __distribute_idle_results(self):
+        # An exception flying out of here probably means a connection
+        # interruption during idle. This will just show like any other
+        # unhandled task exception and that's probably the best we can do.
+        while True:
+            result = await self.__idle_results.get()
+            idle_changes = list(await result)
+            if not idle_changes:
+                continue
+            for subsystems, callback in self.__idle_consumers:
+                if not subsystems or any(s in subsystems for s in idle_changes):
+                    callback(idle_changes)
+
+    # helper methods
+
+    async def __readline(self):
+        """Wrapper around .__rfile.readline that handles encoding"""
+        data = await self.__rfile.readline()
+        try:
+            return data.decode('utf8')
+        except UnicodeDecodeError:
+            self.disconnect()
+            raise ProtocolError("Invalid UTF8 received")
+
+    def __write(self, text):
+        """Wrapper around .__wfile.write that handles encoding."""
+        self.__wfile.write(text.encode('utf8'))
+
+    # copied and subtly modifiedstuff from base
+
+    # This is just a wrapper for the below.
+    def _write_line(self, text):
+        self.__write(text + "\n")
+    # FIXME This code should be shareable.
+    _write_command = SyncMPDClient._write_command
+
+
+    async def __read_output_line(self):
+        """Kind of like SyncMPDClient._read_line"""
+        line = await self.__readline()
+        if not line.endswith("\n"):
+            raise ConnectionError("Connection lost while reading line")
+        line = line.rstrip("\n")
+        if line.startswith(ERROR_PREFIX):
+            error = line[len(ERROR_PREFIX):].strip()
+            raise CommandError(error)
+        if line == SUCCESS:
+            return None
+        return line
+
+
+#    async def _parse_objects_direct(self, lines, delimiters=[]):
```

Comment:
What's up with this version of ``_parse_objects_direct``? Can it be removed?

Reply:
This is the new syntax, that can be eventually used after python 3.5 support have been dropped (see comments below).